contracts/XINV.sol

```diff
@@ -0,0 +1,1190 @@
+     */
+    function escrow(address user, uint amount) public {
+        require(msg.sender == market, "only market can escrow");
+        if(duration > 0) {
+            EscrowData memory withdrawal = pendingWithdrawals[user];
+            pendingWithdrawals[user] = EscrowData({
+                // we set the future withdrawal timestamp based on current `duration` to avoid applying future `duration` changes to existing withdrawals in the event of a governance attack
+                withdrawalTimestamp: block.timestamp + duration,
+                amount: withdrawal.amount.add(amount)
+            });
+            emit Escrow(user, block.timestamp + duration, amount);
+        } else { // if duration is 0, we send the funds directly to the user
+            EIP20Interface token = EIP20Interface(underlying);
+            token.transfer(user, amount);
+        }
+    }
+
+    /**
+     * @notice returns user withdrawable amount
+     */
+    function withdrawable(address user) public view returns (uint amount) {
+        EscrowData memory withdrawal = pendingWithdrawals[user];
+        if(withdrawal.withdrawalTimestamp <= block.timestamp) {
+            amount = withdrawal.amount;
+        }
+    }
+
+    function withdraw() public {
+        uint amount = withdrawable(msg.sender);
+        EscrowData memory withdrawal = pendingWithdrawals[msg.sender];
+        require(amount > 0, "Nothing to withdraw");
+        EIP20Interface token = EIP20Interface(underlying);
+        delete pendingWithdrawals[msg.sender];
+        token.transfer(msg.sender, amount);
+        emit Withdraw(msg.sender, amount);
+    }
+
+    event Escrow(address to, uint withdrawalTimestamp, uint amount);
+    event Withdraw(address to, uint amount);
+}
+
+/**
+ * @title xINV contract
+ * @notice wraps INV token
+ * @author Inverse Finance
+ */
+contract XINV is xInvCore {
+
+    address public underlying;
+    TimelockEscrow public escrow;
+
+    /**
+     * @notice Construct the xINV market
+     * @param underlying_ The address of the underlying asset
+     * @param comptroller_ The address of the Comptroller
+     * @param name_ ERC-20 name of this token
+     * @param symbol_ ERC-20 symbol of this token
+     * @param decimals_ ERC-20 decimal precision of this token
+     * @param admin_ Address of the administrator of this token
+     */
+    constructor(address underlying_,
+                ComptrollerInterface comptroller_,
+                uint rewardPerBlock_,        
+                address rewardTreasury_,
+                string memory name_,
+                string memory symbol_,
+                uint8 decimals_,
+                address payable admin_) public {
+        // Creator of the contract is admin during initialization
+        admin = msg.sender;
+
+        // CToken initialize does the bulk of the work
+        super.initialize(comptroller_, 1e18, rewardPerBlock_, rewardTreasury_, name_, symbol_, decimals_);
```

Comment:
All cTokens are 8 decimals, have we fully considered the implications of having 18 decimals?

Reply:
I've tested it on testnet. Unit tests also pass.